### PR TITLE
[Inductor][ROCm][CK] add CK grouped conv2d fwd kernels to ROCm codegen

### PR DESCRIPTION
**Pull request by tenpercent**
Plug into lowering and end to end test in a later PR

Instance parsing companion PR https://github.com/ROCm/composable_kernel/pull/1585

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @zjing14

### CODE REVIEW

**Comment by chenyang78** [comment]
Thanks, Max! Could we add some tests?

**Comment by tenpercent** [comment]
> Thanks, Max! Could we add some tests?

Could we add the tests later, along with lowering?

**Comment by tenpercent** [comment]
Link #125453

File: torch/_inductor/kernel/mm.py [hunk 1]
```diff
@@ -204,7 +204,11 @@ def tuned_mm(mat1, mat2, *, layout=None):
     if static_shape and is_nonzero and use_cutlass_template(layout, m, n, k):
         CUTLASS3xGemmTemplate.add_cutlass_gemm_choices(choices, layout, [mat1, mat2])
 
-    if is_nonzero and use_ck_template(layout, m, n, k):
+    if (
+        is_nonzero
+        and use_ck_gemm_template(layout)
+        and V.graph.sizevars.size_hint(m * n * k, fallback=-1) > 0
```
Comment (ColinPeppler) [inline torch/_inductor/kernel/mm.py:210]: was thinking whether it made sense to pull this into `use_ck_gemm_template`

**Review by ColinPeppler — APPROVED** [review]

File: torch/_inductor/codegen/rocm/ck_conv_template.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+from torch._inductor.utils import IndentedBuffer
+
+
+log = logging.getLogger(__name__)
+
+
+def torch_layout_to_ck_layouts(torch_layout):
+    # logically, torch tensors are always NCHW,
+    # and channels-last memory layout is visible in the strides
+    if torch_layout.stride[-1] == 1:
```
Comment (ColinPeppler) [inline torch/_inductor/codegen/rocm/ck_conv_template.py:38]: we can use `statically_known_equals(stride[-1], 1)` here for dynamic shape support

https://github.com/pytorch/pytorch/blob/134f6cda7e192046921f44b9565a3b56cd28158d/torch/_inductor/sizevars.py#L340C9-L340C32

File: torch/_inductor/codegen/rocm/ck_conv_template.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+
+    def gen_ops(self):
+        unfiltered_instances = gen_conv_ops_library()
+
+        filtered_instances = list(
+            filter(lambda op: self.filter_op(op), unfiltered_instances)
+        )
+        # NB: when using a fixed list order, most likely we will pick the subset of instances
+        # which are very similar to each other. Randomizing the choice seems to solve this.
+        random.seed(-11)
```
Comment (ColinPeppler) [inline torch/_inductor/codegen/rocm/ck_conv_template.py:467]: also see this in ck_universal_gemm_template, might be nice to reference the same seed here!

https://github.com/pytorch/pytorch/blob/364340c7eef72b840eb61fdd1c52166b93820fdf/torch/_inductor/codegen/rocm/ck_universal_gemm_template.py#L421

Comment (tenpercent) [reply torch/_inductor/codegen/rocm/ck_conv_template.py:469]: This is a copy/paste artifact, I might change it if the instances picked for test do not serve well for testing purpose - that is, when we introduce the end-to-end test

**Review by chenyang78 — APPROVED** [review]

**Pull request closed without merging**